# Exploratory Data Analysis of Spotify Music Tracks

## Introduction

**Spotify** is a digital music streaming service that was founded in Stockholm, Sweden in 2006 by Daniel Ek and Martin Lorentzon. It started as a small startup with the goal of providing a legal and accessible platform for music consumption. Over the years, Spotify has grown into one of the most popular music streaming services worldwide, with millions of active users and an extensive music library.

## Table of Contents

- Import Necessary Libraries
- Data Understanding
- Data Preparation
- Exploratory Data Analysis
- Conclusions

## Import Necessary Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
%matplotlib inline

sns.set_style("darkgrid")

In [2]:
# Read data in csv format

df = pd.read_csv("spotifyfeatures.csv")

## Data Understanding

- head and tail

- sample

- Dataframe shape

- size

- columns

- dtypes

- info

- describe



In [3]:
# preview the first 5 rows of dataset

df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

In [4]:
# preview the last 5 rows of dataset

df.tail()

genre               artist_name                      track_name  \
232720  Soul                     Slave                    Son Of Slide   
232721  Soul  Jr Thomas & The Volcanos                    Burning Fire   
232722  Soul              Muddy Waters  (I'm Your) Hoochie Coochie Man   
232723  Soul                   R.LUM.R                   With My Words   
232724  Soul            Mint Condition  You Don't Have To Hurt No More   

                      track_id  popularity  acousticness  danceability  \
232720  2XGLdVl7lGeq8ksM6Al7jT          39       0.00384         0.687   
232721  1qWZdkBl4UVPj9lK6HuuFM          38       0.03290         0.785   
232722  2ziWXUmQLrXTiYjCg2fZ2t          47       0.90100         0.517   
232723  6EFsue2YbIG4Qkq8Zr9Rir          44       0.26200         0.745   
232724  34XO9RwPMKjbvRry54QzWn          35       0.09730         0.758   

        duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
232720       326240   0.714          0.544000   D    0.0845   -10.626  Major   
232721       282447   0.683          0.000880   E    0.2370    -6.944  Minor   
232722       166960   0.419          0.000000   D    0.0945    -8.282  Major   
232723       222442   0.704          0.000000   A    0.3330    -7.137  Major   
232724       323027   0.470          0.000049  G#    0.0836    -6.708  Minor   

        speechiness    tempo time_signature  valence  
232720       0.0316  115.542            4/4    0.962  
232721       0.0337  113.830            4/4    0.969  
232722       0.1480   84.135            4/4    0.813  
232723       0.1460  100.031            4/4    0.489  
232724       0.0287  113.897            4/4    0.479

In [5]:
# to see any random rows to ensure about the data quality

df.sample(7, random_state=12345)

genre        artist_name                   track_name  \
27447        Alternative            Phoenix                Fior di Latte   
97011   Children’s Music        Xavier Omär              If This Is Love   
211858            Comedy        Doug Benson                   Anti-Truth   
41304               Folk  Car Seat Headrest  Drunk Drivers/Killer Whales   
13642              Dance               Zedd     Stay (with Alessia Cara)   
194412             Movie       Fabien Nataf                        Lilly   
90479            Hip-Hop        Dom Kennedy                   I Love Dom   

                      track_id  popularity  acousticness  danceability  \
27447   1Bit6aQpVL9mQ2jIz4wQO7          40         0.203         0.568   
97011   2v9TxPdblHMBRir1y9uIqw          56         0.224         0.744   
211858  5m7AA94zU702nhxwGEqKL0          10         0.841         0.687   
41304   2os0aK782bakCPmjow0SU0          60         0.218         0.534   
13642   6uBhi9gBXWjanegOb2Phh0          82         0.253         0.690   
194412  76ykXT0nshkQGutCsNK3fv           8         0.372         0.697   
90479   1kFQ2V8nH923JPFzMsNWxS          49         0.278         0.873   

        duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
27447        243733   0.499          0.000532  A#    0.2280    -9.108  Major   
97011        220301   0.472          0.000215   G    0.2250    -9.757  Minor   
211858       105594   0.587          0.000000   B    0.8160    -9.076  Major   
41304        374653   0.414          0.000002   D    0.1030    -7.952  Major   
13642        210091   0.622          0.000000   F    0.1160    -5.025  Minor   
194412       102556   0.799          0.005510   E    0.0811    -4.598  Major   
90479        199674   0.522          0.000000   D    0.0919    -5.801  Major   

        speechiness    tempo time_signature  valence  
27447        0.0307  130.029            4/4    0.628  
97011        0.1110  104.063            4/4    0.281  
211858       0.8750   58.137            4/4    0.275  
41304        0.0403  117.079            4/4    0.458  
13642        0.0622  102.040            4/4    0.544  
194412       0.0540  144.946            4/4    0.961  
90479        0.0565   86.998            4/4    0.535

In [6]:
# shape of dataset

df.shape

(232725, 18)

In [7]:
# all columns of dataset

df.columns

Index(['genre', 'artist_name', 'track_name', 'track_id', 'popularity',
       'acousticness', 'danceability', 'duration_ms', 'energy',
       'instrumentalness', 'key', 'liveness', 'loudness', 'mode',
       'speechiness', 'tempo', 'time_signature', 'valence'],
      dtype='object')

In [8]:
# the summary of the dataset

df.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232725 entries, 0 to 232724
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   genre             232725 non-null  object 
 1   artist_name       232725 non-null  object 
 2   track_name        232725 non-null  object 
 3   track_id          232725 non-null  object 
 4   popularity        232725 non-null  int64  
 5   acousticness      232725 non-null  float64
 6   danceability      232725 non-null  float64
 7   duration_ms       232725 non-null  int64  
 8   energy            232725 non-null  float64
 9   instrumentalness  232725 non-null  float64
 10  key               232725 non-null  object 
 11  liveness          232725 non-null  float64
 12  loudness          232725 non-null  float64
 13  mode              232725 non-null  object 
 14  speechiness       232725 non-null  float64
 15  tempo             232725 non-null  float64
 16  time_signature    23

In [10]:
# the descriptive statistics summary

df.describe().T.round(2)

count       mean        std       min        25%  \
popularity        232725.0      41.13      18.19      0.00      29.00   
acousticness      232725.0       0.37       0.35      0.00       0.04   
danceability      232725.0       0.55       0.19      0.06       0.44   
duration_ms       232725.0  235122.34  118935.91  15387.00  182857.00   
energy            232725.0       0.57       0.26      0.00       0.38   
instrumentalness  232725.0       0.15       0.30      0.00       0.00   
liveness          232725.0       0.22       0.20      0.01       0.10   
loudness          232725.0      -9.57       6.00    -52.46     -11.77   
speechiness       232725.0       0.12       0.19      0.02       0.04   
tempo             232725.0     117.67      30.90     30.38      92.96   
valence           232725.0       0.45       0.26      0.00       0.24   

                        50%        75%         max  
popularity            43.00      55.00      100.00  
acousticness           0.23       0.72        1.00  
danceability           0.57       0.69        0.99  
duration_ms       220427.00  265768.00  5552917.00  
energy                 0.60       0.79        1.00  
instrumentalness       0.00       0.04        1.00  
liveness               0.13       0.26        1.00  
loudness              -7.76      -5.50        3.74  
speechiness            0.05       0.10        0.97  
tempo                115.78     139.05      242.90  
valence                0.44       0.66        1.00

**Popularity**: The popularity of tracks in the dataset ranges from 0 to 100, with a mean value of 41.13. This indicates that there is a diverse range of popularity among the tracks.

**Acousticness**: The acousticness of tracks in the dataset ranges from 0 to 1, with a mean value of 0.37. This suggests that there is variation in the degree of acousticness among the tracks.

**Danceability**: The danceability of tracks in the dataset ranges from 0.06 to 0.99, with a mean value of 0.55. This indicates that there is a range of danceability levels among the tracks.

**Duration**: The duration of tracks in the dataset ranges from 15,387 milliseconds to 5,552,917 milliseconds, with a mean value of approximately 235,122.34 milliseconds. This suggests that there is a wide range of track durations.

**Energy**: The energy of tracks in the dataset ranges from 0 to 1, with a mean value of 0.57. This indicates that there is variation in the energy levels among the tracks.

**Instrumentalness**: The instrumentalness of tracks in the dataset ranges from 0 to 1, with a mean value of 0.15. This suggests that there is variation in the degree of instrumentalness among the tracks.

**Liveness**: The liveness of tracks in the dataset ranges from 0.01 to 1, with a mean value of 0.22. This indicates that there is variation in the level of liveness among the tracks.

**Loudness**: The loudness of tracks in the dataset ranges from -52.46 dB to 3.74 dB, with a mean value of -9.57 dB. This suggests that there is variation in the loudness levels among the tracks.

**Speechiness**: The speechiness of tracks in the dataset ranges from 0.02 to 0.97, with a mean value of 0.12. This indicates that there is variation in the level of speechiness among the tracks.

**Tempo**: The tempo of tracks in the dataset ranges from 30.38 BPM to 242.90 BPM, with a mean value of 117.67 BPM. This suggests that there is a wide range of tempos among the tracks.

**Valence**: The valence of tracks in the dataset ranges from 0 to 1, with a mean value of 0.45. This indicates that there is variation in the emotional positivity among the tracks.

## Data Preparation

- Dropping irrelevant columns and rows
- Identifying missing values
- Identifying duplicated columns
- Renaming columns
- Feature Creation

In [11]:
# the number of missing values

df.isna().sum().sort_values(ascending=False)

genre               0
artist_name         0
time_signature      0
tempo               0
speechiness         0
mode                0
loudness            0
liveness            0
key                 0
instrumentalness    0
energy              0
duration_ms         0
danceability        0
acousticness        0
popularity          0
track_id            0
track_name          0
valence             0
dtype: int64

There is no missing values

In [12]:
# the number of duplicate rows of data

df.duplicated().sum()

0

There is no duplicate rows of data.

In [13]:
# To convert the duration from milliseconds to minutes

df['duration_minutes'] = df['duration_ms'] / 60000

In [14]:
df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  duration_minutes  
0       0.0525  166.969            4/4    0.814          1.656217  
1       0.0868  174.003            4/4    0.816          2.289550  
2       0.0362   99.488            5/4    0.368          2.837783  
3       0.0395  171.758            4/4    0.227          2.540450  
4       0.0456  140.576            4/4    0.390          1.377083

## Exploratory Data Analysis

**Genre Analysis**:

* Plot a bar chart showing the distribution of tracks across different genres.

* Create a horizontal bar chart to visualize the top genres based on the number of tracks.

* Generate a box plot to compare the popularity distribution across different genres.

**Artist Analysis**:

* Create a bar chart to visualize the top artists based on the number of tracks.
* Generate a histogram or violin plot to analyze the popularity distribution for tracks by different artists.
* Create scatter plots to explore relationships between artist attributes (e.g., energy or danceability) and track popularity.

**Track Analysis**:

* Plot a histogram or density plot to analyze the distribution of track durations.
* Create a scatter plot to visualize the relationship between track duration and popularity.
* Generate violin plots to compare the distribution of acousticness or instrumentalness across different track names.

**Popularity Analysis**:

* Create a histogram or density plot to visualize the distribution of track popularity.
* Generate a box plot to compare the popularity distribution across different genres or artists.
* Create a scatter plot to explore relationships between popularity and other attributes (e.g., energy or danceability).

**Attribute Analysis**:

* Plot histograms or density plots to analyze the distribution of different attributes (acousticness, danceability, energy, etc.).
* Create scatter plots or heatmaps to explore relationships or correlations between different attributes.
* Generate grouped bar charts to compare attribute distributions across different genres or artists.
Key and Mode Analysis:

* Create bar charts to visualize the distribution of keys and modes in the dataset.
* Generate box plots or violin plots to compare attribute distributions across different keys or modes.
* Create scatter plots or heatmaps to explore relationships between key or mode and other attributes.

**Time Signature Analysis**:

* Plot a bar chart to visualize the distribution of time signatures in the dataset.
* Create box plots or violin plots to compare attribute distributions across different time signatures.
* Generate scatter plots or heatmaps to explore relationships between time signature and other attributes.

**Valence Analysis**:

* Plot a histogram or density plot to analyze the distribution of valence across tracks.
* Create box plots or violin plots to compare the valence distribution across different genres or artists.
* Generate scatter plots to explore relationships between valence and other attributes (e.g., energy or tempo).

## Conclusions



**Author**: **Ergyun Hasan**  **2023**